Add OTNO, and an example of ShapeNet-Car dataset

Xinyi-Li-4869 opened the pull request:
This PR introduces optimal transport neural operator (OTNO) with an example of ShapeNet-Car dataset.

The main classes are:
1.  OTDataModule: generate OT data from physical mesh
2. OTNO: model implementation

For car dataset example, I added files according to GINO. The main files are:
- neuraloperator/config/otno_carcfd_config.py
- neuraloperator/neuralop/data/datasets/car_ot_dataset.py
- neuraloperator/scripts/train_otno_carcfd.py

Paper link: https://arxiv.org/abs/2507.20065
Original code: https://github.com/Xinyi-Li-4869/OTNO

Xinyi-Li-4869 commented on neuralop/models/otno.py:
```diff
@@ -0,0 +1,101 @@
+import torch.nn as nn
+
+from neuralop.models import FNO
+from neuralop.layers.channel_mlp import ChannelMLP as NeuralopMLP
```
Thanks, have revised to keep the original ChannelMLP name.

Xinyi-Li-4869 commented on scripts/run_test.py:
```diff
@@ -0,0 +1,16 @@
+            root_dir="D:/python_code/data/car-pressure-data",
+            n_train=2,
+            n_test=1,
+        )
```
Have removed

Xinyi-Li-4869 commented on scripts/run_test.py:
```diff
@@ -0,0 +1,16 @@
+
+if __name__ == "__main__":
+    data_module = CarOTDataset(
+            root_dir="D:/python_code/data/car-pressure-data",
```
Thanks, have revised

Xinyi-Li-4869 commented on neuralop/data/datasets/car_ot_dataset.py:
```diff
@@ -0,0 +1,153 @@
+    expand_factor : float, optional
+        Scale factor to map physical mesh size to latent mesh size (e.g., torus/sphere).
+    reg : float, optional
+        Regularization coefficient for the Sinkhorn algorithm.
```
Have added details

Xinyi-Li-4869 commented on neuralop/data/datasets/ot_datamodule.py:
```diff
@@ -0,0 +1,188 @@
+        expand_factor : float, optional
+            Scale factor to map physical mesh size to latent mesh size (e.g., torus/sphere).
+        reg : float, optional
+            Regularization coefficient for the Sinkhorn algorithm.
```
Have added

Xinyi-Li-4869 commented on neuralop/data/datasets/data/ot_expand3.0_reg1e-06_num3.pt:
A more descriptive name would definitely be better — I just don’t have a good idea at the moment. Thanks for the suggestion!

vduruiss commented on examples/models/car_to_torus.gif:
Let's remove the .gif, it should be generated automatically for the tutorial

vduruiss commented on neuralop/models/tests/test_otno.py:
```diff
@@ -0,0 +1,76 @@
+@pytest.mark.parametrize("factorization", [None, "tucker"])
+@pytest.mark.parametrize("n_layers", [2, 4])
+@pytest.mark.parametrize("positional_embedding", [None, "grid"]) 
+@pytest.mark.parametrize("domain_padding", [0.0, 0.2, 0.4])
```
This is good, but it will run the code 144 times to test every single possible combination.

Can we break this into 2-3 test functions? This way it runs less combinations and would not require as much time to run for test/deployment

vduruiss commented on doc/requirements_doc.txt:
```diff
@@ -5,4 +5,10 @@ sphinx
+imageio
+imageio[ffmpeg]
+imageio[pyav]
+pillow
```
See the comments for the tutorial. Maybe we can have the animations show up without requiring all these modules

Xinyi-Li-4869 commented:
I’ve removed the GIFs and rewritten the content following the animation example, so I also removed the related packages from the documentation environment requirements. I’ve also reduced the test functions.

Please check again @vduruiss  — thanks so much!

JeanKossaifi commented:
Looks good to me as well, thank you both @Xinyi-Li-4869 @vduruiss !
